### ISSUE TITLE
Issue : Error adding package error loading bundle into db: prometheusoperator.0.22.2 specifies replacement that couldn't be found

### ISSUE THREAD

**Issue by cmoulliard** [operator-framework/operator-registry#178]
## Issue

I build using the master branch - 0652a72 of this project the `olm` tool and created a bundle that I pushed on quay

```
./opm alpha bundle build -d manifests/prometheus/0.22.2 --tag quay.io/cmoulliard/olm-prometheus:0.22.2 -p prometheus -c preview -e preview
INFO[0000] Building annotations.yaml                    
INFO[0000] An annotations.yaml already exists in directory 
INFO[0000] Validating existing annotations.yaml         
INFO[0000] Building Dockerfile                          
INFO[0000] Building bundle image                        
Sending build context to Docker daemon  359.9kB
Step 1/9 : FROM scratch
 ---> 
Step 2/9 : LABEL operators.operatorframework.io.bundle.mediatype.v1=registry+v1
 ---> Using cache
 ---> 9a1a43532027
Step 3/9 : LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
 ---> Using cache
 ---> 1e7a1f544437
Step 4/9 : LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
 ---> Using cache
 ---> e1d6fa13ad0f
Step 5/9 : LABEL operators.operatorframework.io.bundle.package.v1=prometheus
 ---> Using cache
 ---> 5dca8ee60ad4
Step 6/9 : LABEL operators.operatorframework.io.bundle.channels.v1=preview
 ---> Using cache
 ---> 2bc58968e68c
Step 7/9 : LABEL operators.operatorframework.io.bundle.channel.default.v1=preview
 ---> Using cache
 ---> fa74f3581eef
Step 8/9 : COPY /*.yaml /manifests/
 ---> Using cache
 ---> da1536f4ea60
Step 9/9 : COPY /metadata/annotations.yaml /metadata/annotations.yaml
 ---> Using cache
 ---> f0151117de1b
Successfully built f0151117de1b
Successfully tagged quay.io/cmoulliard/olm-prometheus:0.22.2

docker push  quay.io/cmoulliard/olm-prometheus:0.22.2
The push refers to repository [quay.io/cmoulliard/olm-prometheus]
3b4589df344b: Pushed 
2e27d9292f04: Pushed 
0.22.2: digest: sha256:55bb192fbec86d93b35059a4cb279961e32f58d1877ff306f433a0dcb9c9523a size: 733
```

but when I try to add the index to the image `quay.io/cmoulliard/olm-index` , then I get this error 

```
bundleImage=quay.io/cmoulliard/olm-prometheus
./opm index add -b="$bundleImage:0.22.2" -t quay.io/cmoulliard/olm-index:0.1.0 -c=docker
INFO[0000] building the index                            bundles="[quay.io/cmoulliard/olm-prometheus:0.22.2]"
INFO[0000] running docker pull                           img="quay.io/cmoulliard/olm-prometheus:0.22.2"
INFO[0001] running docker save                           img="quay.io/cmoulliard/olm-prometheus:0.22.2"
INFO[0002] loading Bundle quay.io/cmoulliard/olm-prometheus:0.22.2  img="quay.io/cmoulliard/olm-prometheus:0.22.2"
INFO[0002] found annotations file searching for csv      dir=bundle_tmp419156138 file=bundle_tmp419156138/metadata load=annotations
INFO[0002] found csv, loading bundle                     dir=bundle_tmp419156138 file=bundle_tmp419156138/manifests load=bundle
INFO[0002] loading bundle file                           dir=bundle_tmp419156138/manifests file=alertmanager.crd.yaml load=bundle
INFO[0002] loading bundle file                           dir=bundle_tmp419156138/manifests file=prometheus.crd.yaml load=bundle
INFO[0002] loading bundle file                           dir=bundle_tmp419156138/manifests file=prometheusoperator.0.22.2.clusterserviceversion.yaml load=bundle
INFO[0002] loading bundle file                           dir=bundle_tmp419156138/manifests file=prometheusrule.crd.yaml load=bundle
INFO[0002] loading bundle file                           dir=bundle_tmp419156138/manifests file=servicemonitor.crd.yaml load=bundle
FATA[0002] permissive mode disabled                      bundles="[quay.io/cmoulliard/olm-prometheus:0.22.2]" error="error loading bundle from image: Error adding package error loading bundle into db: prometheusoperator.0.22.2 specifies replacement that couldn't be found"
```

## Info

OS: MacOS
Go version: 1.13.7

**Comment by cmoulliard** [comment]
I can avoid this error if I pass as paramete r`--permissive` but then I get another error

```
ERRO[0220] Sending build context to Docker daemon  143.3MB
Step 1/9 : FROM quay.io/operator-framework/upstream-registry-builder AS builder
latest: Pulling from operator-framework/upstream-registry-builder
c9b1b535fdd9: Pulling fs layer
cbb0d8da1b30: Pulling fs layer
d909eff28200: Pulling fs layer
f9e7c84be0a1: Pulling fs layer
8a0c202c45df: Pulling fs layer
b2533e9c644c: Pulling fs layer
191e0c84c848: Pulling fs layer
4a13ea8d9175: Pulling fs layer
c17eee0ccc2c: Pulling fs layer
8fc84b6e6f4b: Pulling fs layer
174ac6379960: Pulling fs layer
46b65b497c5d: Pulling fs layer
c6baaf243336: Pulling fs layer
37f08974078a: Pulling fs layer
ca97294f2eec: Pulling fs layer
191e0c84c848: Waiting
f9e7c84be0a1: Waiting
8a0c202c45df: Waiting
b2533e9c644c: Waiting
8fc84b6e6f4b: Waiting
174ac6379960: Waiting
46b65b497c5d: Waiting
4a13ea8d9175: Waiting
c17eee0ccc2c: Waiting
c6baaf243336: Waiting
37f08974078a: Waiting
ca97294f2eec: Waiting
d909eff28200: Download complete
cbb0d8da1b30: Verifying Checksum
cbb0d8da1b30: Download complete
8a0c202c45df: Download complete
c9b1b535fdd9: Verifying Checksum
c9b1b535fdd9: Download complete
c9b1b535fdd9: Pull complete
cbb0d8da1b30: Pull complete
d909eff28200: Pull complete
191e0c84c848: Download complete
4a13ea8d9175: Verifying Checksum
4a13ea8d9175: Download complete
c17eee0ccc2c: Verifying Checksum
c17eee0ccc2c: Download complete
8fc84b6e6f4b: Verifying Checksum
8fc84b6e6f4b: Download complete
174ac6379960: Verifying Checksum
174ac6379960: Download complete
46b65b497c5d: Download complete
b2533e9c644c: Download complete
37f08974078a: Verifying Checksum
37f08974078a: Download complete
c6baaf243336: Verifying Checksum
c6baaf243336: Download complete
f9e7c84be0a1: Download complete
f9e7c84be0a1: Pull complete
8a0c202c45df: Pull complete
ca97294f2eec: Verifying Checksum
ca97294f2eec: Download complete
b2533e9c644c: Pull complete
191e0c84c848: Pull complete
4a13ea8d9175: Pull complete
c17eee0ccc2c: Pull complete
8fc84b6e6f4b: Pull complete
174ac6379960: Pull complete
46b65b497c5d: Pull complete
c6baaf243336: Pull complete
37f08974078a: Pull complete
ca97294f2eec: Pull complete
Digest: sha256:7d8e5dddad0275bc903b6ef17840f98441b15b8b999609af2c9579960e52080e
Status: Downloaded newer image for quay.io/operator-framework/upstream-registry-builder:latest
 ---> ce4635d54532
Step 2/9 : FROM scratch
 ---> 
Step 3/9 : LABEL operators.operatorframework.io.index.database.v1=./index.db
 ---> Running in 7eb1c8ae186f
Removing intermediate container 7eb1c8ae186f
 ---> 825eb27b2ae2
Step 4/9 : COPY index_tmp559487510 ./
 ---> 7f368e3c990e
Step 5/9 : COPY --from=builder /build/bin/opm /opm
COPY failed: stat /var/lib/docker/overlay2/967cc6c71ddcccf1541386cf6938c324fd7c18a5f35f4872d874c1d055c5c241/merged/build/bin/opm: no such file or directory  bundles="[quay.io/cmoulliard/olm-prometheus:0.22.2]"
Error: error building image: Sending build context to Docker daemon  143.3MB
Step 1/9 : FROM quay.io/operator-framework/upstream-registry-builder AS builder
latest: Pulling from operator-framework/upstream-registry-builder
c9b1b535fdd9: Pulling fs layer
cbb0d8da1b30: Pulling fs layer
d909eff28200: Pulling fs layer
f9e7c84be0a1: Pulling fs layer
8a0c202c45df: Pulling fs layer
b2533e9c644c: Pulling fs layer
191e0c84c848: Pulling fs layer
4a13ea8d9175: Pulling fs layer
c17eee0ccc2c: Pulling fs layer
8fc84b6e6f4b: Pulling fs layer
174ac6379960: Pulling fs layer
46b65b497c5d: Pulling fs layer
c6baaf243336: Pulling fs layer
37f08974078a: Pulling fs layer
ca97294f2eec: Pulling fs layer
191e0c84c848: Waiting
f9e7c84be0a1: Waiting
8a0c202c45df: Waiting
b2533e9c644c: Waiting
8fc84b6e6f4b: Waiting
174ac6379960: Waiting
46b65b497c5d: Waiting
4a13ea8d9175: Waiting
c17eee0ccc2c: Waiting
c6baaf243336: Waiting
37f08974078a: Waiting
ca97294f2eec: Waiting
d909eff28200: Download complete
cbb0d8da1b30: Verifying Checksum
cbb0d8da1b30: Download complete
8a0c202c45df: Download complete
c9b1b535fdd9: Verifying Checksum
c9b1b535fdd9: Download complete
c9b1b535fdd9: Pull complete
cbb0d8da1b30: Pull complete
d909eff28200: Pull complete
191e0c84c848: Download complete
4a13ea8d9175: Verifying Checksum
4a13ea8d9175: Download complete
c17eee0ccc2c: Verifying Checksum
c17eee0ccc2c: Download complete
8fc84b6e6f4b: Verifying Checksum
8fc84b6e6f4b: Download complete
174ac6379960: Verifying Checksum
174ac6379960: Download complete
46b65b497c5d: Download complete
b2533e9c644c: Download complete
37f08974078a: Verifying Checksum
37f08974078a: Download complete
c6baaf243336: Verifying Checksum
c6baaf243336: Download complete
f9e7c84be0a1: Download complete
f9e7c84be0a1: Pull complete
8a0c202c45df: Pull complete
ca97294f2eec: Verifying Checksum
ca97294f2eec: Download complete
b2533e9c644c: Pull complete
191e0c84c848: Pull complete
4a13ea8d9175: Pull complete
c17eee0ccc2c: Pull complete
8fc84b6e6f4b: Pull complete
174ac6379960: Pull complete
46b65b497c5d: Pull complete
c6baaf243336: Pull complete
37f08974078a: Pull complete
ca97294f2eec: Pull complete
Digest: sha256:7d8e5dddad0275bc903b6ef17840f98441b15b8b999609af2c9579960e52080e
Status: Downloaded newer image for quay.io/operator-framework/upstream-registry-builder:latest
 ---> ce4635d54532
Step 2/9 : FROM scratch
 ---> 
Step 3/9 : LABEL operators.operatorframework.io.index.database.v1=./index.db
 ---> Running in 7eb1c8ae186f
Removing intermediate container 7eb1c8ae186f
 ---> 825eb27b2ae2
Step 4/9 : COPY index_tmp559487510 ./
 ---> 7f368e3c990e
Step 5/9 : COPY --from=builder /build/bin/opm /opm
COPY failed: stat /var/lib/docker/overlay2/967cc6c71ddcccf1541386cf6938c324fd7c18a5f35f4872d874c1d055c5c241/merged/build/bin/opm: no such file or directory
. exit status 1
Usage:
  opm index add [flags]

Examples:
  # Create an index image from scratch with a single bundle image
  opm index add --bundles quay.io/operator-framework/operator-bundle-prometheus@sha256:a3ee653ffa8a0d2bbb2fabb150a94da6e878b6e9eb07defd40dc884effde11a0 --tag quay.io/operator-framework/monitoring:1.0.0
  
  # Add a single bundle image to an index image
  opm index add --bundles quay.io/operator-framework/operator-bundle-prometheus:0.15.0 --from-index quay.io/operator-framework/monitoring:1.0.0 --tag quay.io/operator-framework/monitoring:1.0.1
  
  # Add multiple bundles to an index and generate a Dockerfile instead of an image
  opm index add --bundles quay.io/operator-framework/operator-bundle-prometheus:0.15.0,quay.io/operator-framework/operator-bundle-prometheus:0.22.2 --generate

Flags:
  -i, --binary-image opm        container image for on-image opm command
  -b, --bundles strings         comma separated list of bundles to add
  -c, --container-tool string   tool to interact with container images (save, build, etc.). One of: [docker, podman] (default "podman")
  -f, --from-index string       previous index to add to
      --generate                if enabled, just creates the dockerfile and saves it to local disk
  -h, --help                    help for add
  -d, --out-dockerfile string   if generating the dockerfile, this flag is used to (optionally) specify a dockerfile name
      --permissive              allow registry load errors
  -t, --tag string              custom tag for container image being built
```

Error is 
```
COPY failed: stat /var/lib/docker/overlay2/967cc6c71ddcccf1541386cf6938c324fd7c18a5f35f4872d874c1d055c5c241/merged/build/bin/opm: no such file or directory  bundles="[quay.io/cmoulliard/olm-prometheus:0.22.2]"
```

**Comment by cmoulliard** [comment]
Related to issue #176. Can the fix merged ASAP please ?

**Comment by aelbarkani** [comment]
I get the same error regarding replacement, even if the specified csv seems to be there. Then I used the --permissive flag as indicated by @cmoulliard and seems to be working with opm version 1.5.10. although I couldn't assess what are the real impacts of this workaround

**Comment by aelbarkani** [comment]
actually the workaround doesn't work. it seems opm ignores the new bundle and it doesn't add it to the index

**Comment by aelbarkani** [comment]
@cmoulliard if it helps you, this problem arises when you the bundle to be replaced is not in the same channel than the bundle you are adding to the index. so both bundles should be in the same channel. that solved the problem for me

**Comment by rcernich** [comment]
If it helps, I got a similar error, but the real problem was an invalid csv in a different set of manifests.  Nothing in the reported error was relevant to the actual problem.  (I was creating a registry with three sets of manifests and the csv file that was bad was in a different set of manifests than the csv file reported in the error message.)

**Comment by rcernich** [comment]
Ignore my last comment.  This was just build caching in docker/podman.  While the metadata in the csv was bad, it didn't actually "fix" this problem.

**Comment by ankitathomas** [comment]
the replacement error and the COPY error are unrelated.

The COPY error should be fixed as of v1.5.10, the fix has already merged with https://github.com/operator-framework/operator-registry/pull/173.

"prometheusoperator.0.22.2 specifies replacement that couldn't be found" isn't a bug, this happens because of how opm handles entries in the 'replaces' field in the CSV in replaces mode. Any bundle version mentioned in the replaces field must either be present in the index, or be in the list of bundles to be added.
The CSV for prometheusoperator.0.22.2 https://github.com/operator-framework/operator-registry/blob/master/bundles/prometheus.0.22.2/manifests/prometheusoperator.0.22.2.clusterserviceversion.yaml#L10 has prometheusoperator.0.15.0 in its replaces field, and 0.15.0 has 0.14.0 in its replaces field. You'd need to add all 3 bundles together, or ensure that a replacement was already present.

To add 0.22.2 alone, you'd either have to rebuild the bundle with the previous version in the skips field instead, or just remove the replaces field from the CSV.

https://github.com/operator-framework/operator-lifecycle-manager/blob/7c2bd2ddcb469b5447890b9deca0c293ce0dcfe9/doc/design/how-to-update-operators.md has more on how upgrades work in bundles, and how skips/replaces work.
